### PR TITLE
[Cache] Add FilesystemAdapter

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ */
+class FilesystemAdapter extends AbstractAdapter
+{
+    private $directory;
+
+    public function __construct($directory, $defaultLifetime = null)
+    {
+        parent::__construct('', $defaultLifetime);
+
+        $dir = rtrim($directory, '/'.DIRECTORY_SEPARATOR).DIRECTORY_SEPARATOR;
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: Why do we use the `DIRECTORY_SEPARATOR` constant here (and below) to construct paths?

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        if (!file_exists($dir)) {
+            @mkdir($dir, 0777, true);
+        }
+        if (false === $dir = realpath($dir)) {
+            throw new InvalidArgumentException(sprintf('Cache directory does not exist (%s)', $directory));
+        }
+        if (!is_writable($dir .= DIRECTORY_SEPARATOR)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: `$dir` already ends with `DIRECTORY_SEPARATOR`

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        parent::__construct('', $defaultLifetime);
+
+        $dir = rtrim($directory, '/'.DIRECTORY_SEPARATOR).DIRECTORY_SEPARATOR;
+
+        if (!file_exists($dir)) {
+            @mkdir($dir, 0777, true);
+        }
+        if (false === $dir = realpath($dir)) {
+            throw new InvalidArgumentException(sprintf('Cache directory does not exist (%s)', $directory));
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: I would move `(%s)` directly after `Cache directory` (same with the other messages below).

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            $file = $this->getFile($id);
+            $dir = dirname($file);
+            if (!file_exists($dir)) {
+                @mkdir($dir, 0777, true);
+            }
+            $value = $expiresAt."\n".serialize($value);
+            if (false !== @file_put_contents($file, $value, LOCK_EX)) {
+                @touch($file, $expiresAt);
+            } else {
+                $ok = false;
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: The name of the `$ok` variable makes lines like this confusing to read: `$ok = false`. Could we rename it as `$result`?

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+     * {@inheritdoc}
+     */
+    protected function doFetch(array $ids)
+    {
+        $values = array();
+        $now = time();
+
+        foreach ($ids as $id) {
+            $file = $this->getFile($id);
+            if (!$h = @fopen($file, 'rb')) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: The `$h` variable name could also be replaced for something more meaningful.

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: To "file_exists" a directory, not a file

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: I prefer having the message in front to make it more readable. The patch can be quite long and could push the interesting info out of the viewport or on the next line.

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: realpath can remove it (depends on OS)

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: I thought always using the forward slash works on Winwows too, doesn't it?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php]: It works the same yes. I tweaked the code a bit. ok for you?

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**